File: golang/longest_common_prefix.go
// find the longest common prefix amongst an array of strings
// you should choose the word with the shortest length.
func longestCommonPrefix(strs []string) string {
	prefix := ""

	// better to find the min len among all strings
	minLen := len(strs[0])
	for _, str := range strs {
		if len(str) < minLen {
			minLen = len(str)
		}
	}
	if minLen == 0 {
		return prefix
	}

	for pos := 0; pos < minLen; pos++ {
		char := strs[0][pos]

		for _, str := range strs {
			if str[pos] != char {
				return prefix
			}
		}

		prefix += string(char)

		if pos == minLen-1 {
			return prefix
		}
	}

	return prefix
}

File: typescript/longest_common_prefix.ts
function longestCommonPrefix(strs: string[]): string {
	// find smallest string
	let smallest = strs[0]
	strs.forEach(str => {
		if (str.length < smallest.length) smallest = str
	})

	// build prefix
	let prefix = ''
	for (let i = 0; i < smallest.length; i++) {
		// all strings must qualify to build => any failure -> break
		let success = true
		for (let str of strs) {
			if (str[i] != smallest[i]) {
				success = false
				break
			}
		}

		if (success) prefix += smallest[i]
		else break // found failure => end
	}

	return prefix
}
